### ISSUE TITLE
Unable to share TSWasmEngine between multiple TSWasmStore

### ISSUE THREAD

**Issue by CGamesPlay** [tree-sitter/tree-sitter#3454]
### Problem

While investigating adding wasm support to `py-tree-sitter`, I noticed that it isn't possible to reuse TSWasmStore objects. The goal is to create a WASM store, load a bunch of languages into it, and then parse a bunch of files in those languages.

- **Attempt 1**: Create a store, load all languages into it, create a parser for each language, share the store between all parsers.
  **Problem 1**: The parser assumes ownership over the passed store and frees it in [ts_parser_set_wasm_store](https://github.com/tree-sitter/tree-sitter/blob/3095fbe07bbcc605910b7ad8bbd6313e0a07e678/lib/src/parser.c#L2146) (called from `ts_parser_delete`).
- **Attempt 2**: Create a store, load all languages into it, create a parser for each language, using a new store for each parser (inspired by [this comment](https://github.com/tree-sitter/tree-sitter/blob/3095fbe07bbcc605910b7ad8bbd6313e0a07e678/lib/include/tree_sitter/api.h#L1198)).
  **Problem 2**: The store assumes ownership over its engine and frees it in [ts_wasm_store_delete](https://github.com/tree-sitter/tree-sitter/blob/3095fbe07bbcc605910b7ad8bbd6313e0a07e678/lib/src/wasm_store.c#L921).
- **Attempt 3**: Same as before, but using a separate engine for each parser.
  **Problem 3:** It is [not possible](https://github.com/bytecodealliance/wasmtime/blob/a068bfe06435a048502bf4afc426451dab5294af/crates/wasmtime/src/runtime/instance.rs#L255) to use a type from another engine in a store .
- **Attempt 4**: For each parser, create a separate engine, dedicated store, load all relevant languages into the store.
  **Problem 4**: Non-ergonomic and inefficient.

Now, I don't know the point of sharing a store between languages. I assume it's because embedded languages need to be in the same store, but I'm just speculating. I can certainly see value in have a different store for each parser, to control memory usage. Based on this, I'm going to suggest that the ideal architecture allows parsers, languages, and stores to share the underlying resources with reference counting (this is already how things are implemented [internally in wasmtime](https://github.com/bytecodealliance/wasmtime/blob/a068bfe06435a048502bf4afc426451dab5294af/crates/c-api/include/doc-wasm.h#L20)).

1. In wasmtime, create a C API to expose cloning engine references.
2. Modify `ts_wasm_store_new` clone its passed engine reference.
3. Modify `TSWasmStore` to be a reference counted object.
4. Deprecate `ts_parser_take_wasm_store`; if necessary, replace with a `ts_parser_get_wasm_store` method that returns a new reference.

Would such a set of modifications be approved in a pull request? Also, I'm not sure what the use case is for `ts_parser_take_wasm_store`, so I'd appreciate feedback on that aspect.

### Steps to reproduce

Create a WASM store, load a bunch of languages into it, create a parser for each language, free everything.

### Expected behavior

No crash from double-free.

### Tree-sitter version (tree-sitter --version)

Tree-sitter 0.22.6

### Operating system/version

macos 12.7.1

**Comment by maxbrunsfeld** [comment]
The intent is that wasm stores are pretty much 1:1 with parsers. There’s no reason to try to share wasm stores between parsers. Both objects can only be used by one thread at a time.

**Comment by CGamesPlay** [comment]
Hi @maxbrunsfeld, thanks for the reply!

I've read the code a bit more thoroughly after reading your note, and if I'm understanding correctly, when I call `ts_parser_set_language`, it [completely copies](https://github.com/tree-sitter/tree-sitter/blob/3095fbe07bbcc605910b7ad8bbd6313e0a07e678/lib/src/wasm_store.c#L1437) the language into the Parser's store, meaning that the store used to load the Language is no longer used at all for parsing. So the intended usage for multiple stores is to use 1-N WasmStores to contain the Language, and then give each Parser a separate WasmStore.

If this is correct, then it looks like the only problem is the WasmStore assuming ownership of the WasmEngine. So my approach will be as follows:
1. Land [bytecodealliance/wasmtime#8907](https://github.com/bytecodealliance/wasmtime/pull/8907), to allow cloning a `wasm_engine_t`.
2. Modify `ts_wasm_store_new` clone its passed engine reference.

Further, when it comes to Python, I imagine that WasmStore won't appear in the public API at all, but I'll open a PR to discuss that when it's ready.

**Comment by maxbrunsfeld** [comment]
The wasmtime engine is already a ref-counted, shareable object, so normally it is shared between all wasm stores.

**Comment by CGamesPlay** [comment]
Yeah, except TSWasmStore assumes ownership over its engine and frees it in [ts_wasm_store_delete](https://github.com/tree-sitter/tree-sitter/blob/3095fbe07bbcc605910b7ad8bbd6313e0a07e678/lib/src/wasm_store.c#L921). Since there isn’t currently any way to increase the ref count from the C API, this will always deallocate the engine in the current implementation.

**Comment by maxbrunsfeld** [comment]
The wasm engine has ref counting APIs provided by wasmtime.

**Comment by CGamesPlay** [comment]
Here's a quick segfault. This is fixed on my development branch which is based off https://github.com/bytecodealliance/wasmtime/pull/8907:

```c
    TSWasmError error;
    TSWasmEngine *engine = wasm_engine_new();
    TSWasmStore *store_a = ts_wasm_store_new(engine, &error);
    TSWasmStore *store_b = ts_wasm_store_new(engine, &error);
    ts_wasm_store_delete(store_a);
    ts_wasm_store_delete(store_b); // Segfault in wasm_engine_delete
    wasm_engine_delete(engine); // This would have the same segfault if you comment out store_b
```

**Comment by maxbrunsfeld** [comment]
Yes, `ts_wasm_store_new` takes ownership of the engine. The ownership model is modeled after the Rust API. If you want to use an engine again, you can use [`wasmtime_engine_clone`](https://docs.wasmtime.dev/c-api/engine_8h.html#a52f38d554bf8146dfb1621b60f15c93f), which bumps the reference count.

**Comment by CGamesPlay** [comment]
I appreciate your comment, that’s literally the method I added in the pull request I mentioned 🙂

**Comment by maxbrunsfeld** [comment]
Oh ha, well thanks for adding that to wasmtime. I didn’t realize they didn’t have C APIs for that.

**Comment by CGamesPlay** [comment]
I'd like to push back on that, if I may.

> this is working as intended (caller should do the cloning), like in Rust

1. The caller typically does not do the cloning when using `wasmtime::Engine` objects. See: [wasmtime::Module::new](https://docs.wasmtime.dev/api/wasmtime/struct.Module.html#method.new), [wasmtime::Store::new](https://docs.wasmtime.dev/api/wasmtime/struct.Store.html#method.new). Wasmtime automatically handles the reference counts.
2. The caller does not do the cloning when calling `ts_parser_set_language`. Tree-sitter automatically handles the reference counts.
3. Pushing reposibility on the caller means that failure to properly clone results in a use-after-free. If the API handles the cloning internally, the worst that can happen is a memory leak.

In fact, I am betting that the only reason that the current rust binding call clone themselves is because it wasn't possible to clone references in C when the code was written originally.

I've currently got a PR out (https://github.com/bytecodealliance/wasmtime/pull/9031) which will unblock tree-sitter upgrading to wasmtime 23 (https://github.com/tree-sitter/tree-sitter/pull/3428). After that point, I'd like to submit another PR that resolves this issue and changes all Rust APIs to accept `&wasmtime::Engine` instead of owned `wasmtime::Engine`.

**Comment by maxbrunsfeld** [comment]
Ok, that seems fine. Good point about consistency with `ts_parser_set_language`.
